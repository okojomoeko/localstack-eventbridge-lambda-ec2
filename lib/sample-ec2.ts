import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export class SampleEC2 extends Construct {
  public readonly instance: ec2.Instance;
  public readonly vpc: ec2.Vpc;
  constructor(scope: Construct, id: string) {
    super(scope, id);
    // わかりやすくAZは1つで今回試す
    // localstackのfree版の都合で、Private Subnet配置のEC2については対象外
    this.vpc = new ec2.Vpc(this, 'VPC', {
      maxAzs: 1,
      subnetConfiguration: [
        {
          subnetType: ec2.SubnetType.PUBLIC,
          name: 'PublicSubnet',
        },
      ],
    });

    this.instance = new ec2.Instance(this, 'SampleEC2', {
      vpc: this.vpc,
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.NANO,
      ),
      machineImage: new ec2.AmazonLinuxImage({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
      }),
    });
  }
}
