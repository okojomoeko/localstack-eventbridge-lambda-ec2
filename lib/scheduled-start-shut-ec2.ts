import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import { Construct } from 'constructs';

export interface ScheduledStartStopEC2Props {
  /**
   * The EC2 instance ID for scheduled start and stop
   */
  ec2InstanceId: string;
  vpc: ec2.Vpc;
}

export class ScheduledStartStopEC2 extends Construct {
  constructor(scope: Construct, id: string, props: ScheduledStartStopEC2Props) {
    super(scope, id);
    const startEC2Lambda = new lambda.Function(this, 'StartEC2Lambda', {
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'start_ec2.lambda_handler',
      code: lambda.Code.fromAsset('./lib/lambda'),
      environment: {
        EC2_INSTANCE_ID: props.ec2InstanceId,
      },
      // 今回はPrivate SubnetにLambdaを配置しない構成
      // vpc: props.vpc,
      // vpcSubnets: props.vpc.selectSubnets({
      //   subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      // }),
    });
    startEC2Lambda.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['ec2:Start*'],
        resources: ['*'], // 面倒なのでひとまずwildcard
      }),
    );

    const stopEC2Lambda = new lambda.Function(this, 'StopEC2Lambda', {
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'stop_ec2.lambda_handler',
      code: lambda.Code.fromAsset('./lib/lambda'),
      environment: {
        EC2_INSTANCE_ID: props.ec2InstanceId,
      },
      // 今回はPrivate SubnetにLambdaを配置しない構成
      // vpc: props.vpc,
      // vpcSubnets: props.vpc.selectSubnets({
      //   subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      // }),
    });
    stopEC2Lambda.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['ec2:Stop*'],
        resources: ['*'], // 面倒なのでひとまずwildcard
      }),
    );

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const startEvent = new events.Rule(this, 'EC2StartEvent', {
      // すぐに確認したいので適当に偶数分ごとに実行する
      schedule: events.Schedule.cron({ minute: '0/2' }),
      targets: [new targets.LambdaFunction(startEC2Lambda)],
    });

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const stopEvent = new events.Rule(this, 'EC2StopEvent', {
      // すぐに確認したいので適当に奇数分ごとに実行する
      schedule: events.Schedule.cron({ minute: '1/2' }),
      targets: [new targets.LambdaFunction(stopEC2Lambda)],
    });

    // localstack Proじゃないと、EventBridgeからPrivate Subnet に配置したLambdaを叩くためのvpc endpointが作れないので保留
    // props.vpc.addInterfaceEndpoint('EventBridgeEndpoint', {
    //   service: ec2.InterfaceVpcEndpointAwsService.EVENTBRIDGE,
    // });
  }
}
