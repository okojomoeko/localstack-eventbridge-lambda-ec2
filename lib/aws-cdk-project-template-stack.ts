import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SampleEC2 } from './sample-ec2';
import { ScheduledStartStopEC2 } from './scheduled-start-shut-ec2';

export class AwsCdkProjectTemplateStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const sampleEc2 = new SampleEC2(this, 'SampleEC2');
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const scheduledStartStopEC2 = new ScheduledStartStopEC2(
      this,
      'ScheduledStartStopEC2',
      {
        ec2InstanceId: sampleEc2.instance.instanceId,
        vpc: sampleEc2.vpc,
      },
    );
  }
}
