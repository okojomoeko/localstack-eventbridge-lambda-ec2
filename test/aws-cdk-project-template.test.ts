import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as AwsCdkProjectTemplate from '../lib/aws-cdk-project-template-stack';

test('Snapshot test', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new AwsCdkProjectTemplate.AwsCdkProjectTemplateStack(
    app,
    'MyTestStack',
  );
  // THEN
  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});
