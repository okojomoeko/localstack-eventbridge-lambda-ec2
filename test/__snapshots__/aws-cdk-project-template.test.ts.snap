// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "SampleEC243038771": {
      "DependsOn": [
        "SampleEC2InstanceRole349DA0CB",
      ],
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "IamInstanceProfile": {
          "Ref": "SampleEC2InstanceProfile59E07FDB",
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "SampleEC2InstanceSecurityGroup839B849B",
              "GroupId",
            ],
          },
        ],
        "SubnetId": {
          "Ref": "SampleEC2VPCPublicSubnetSubnet1Subnet651716FD",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/SampleEC2",
          },
        ],
        "UserData": {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "SampleEC2InstanceProfile59E07FDB": {
      "Properties": {
        "Roles": [
          {
            "Ref": "SampleEC2InstanceRole349DA0CB",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "SampleEC2InstanceRole349DA0CB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/SampleEC2",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SampleEC2InstanceSecurityGroup839B849B": {
      "Properties": {
        "GroupDescription": "MyTestStack/SampleEC2/SampleEC2/InstanceSecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/SampleEC2",
          },
        ],
        "VpcId": {
          "Ref": "SampleEC2VPCF44F6120",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "SampleEC2VPCF44F6120": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "SampleEC2VPCIGWA2B7A9F3": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "SampleEC2VPCPublicSubnetSubnet1DefaultRouteD64FB370": {
      "DependsOn": [
        "SampleEC2VPCVPCGW87075A7C",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "SampleEC2VPCIGWA2B7A9F3",
        },
        "RouteTableId": {
          "Ref": "SampleEC2VPCPublicSubnetSubnet1RouteTable57E376BC",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "SampleEC2VPCPublicSubnetSubnet1RouteTable57E376BC": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/VPC/PublicSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "SampleEC2VPCF44F6120",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "SampleEC2VPCPublicSubnetSubnet1RouteTableAssociation0A0BE4EF": {
      "Properties": {
        "RouteTableId": {
          "Ref": "SampleEC2VPCPublicSubnetSubnet1RouteTable57E376BC",
        },
        "SubnetId": {
          "Ref": "SampleEC2VPCPublicSubnetSubnet1Subnet651716FD",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "SampleEC2VPCPublicSubnetSubnet1Subnet651716FD": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/16",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "PublicSubnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/SampleEC2/VPC/PublicSubnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "SampleEC2VPCF44F6120",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "SampleEC2VPCVPCGW87075A7C": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "SampleEC2VPCIGWA2B7A9F3",
        },
        "VpcId": {
          "Ref": "SampleEC2VPCF44F6120",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "ScheduledStartStopEC2EC2StartEvent2518C7CF": {
      "Properties": {
        "ScheduleExpression": "cron(0/2 * * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ScheduledStartStopEC2StartEC2Lambda896098AC",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduledStartStopEC2EC2StartEventAllowEventRuleMyTestStackScheduledStartStopEC2StartEC2Lambda8357FB5330670E68": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2StartEC2Lambda896098AC",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2EC2StartEvent2518C7CF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduledStartStopEC2EC2StopEvent1CF6ECA1": {
      "Properties": {
        "ScheduleExpression": "cron(1/2 * * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ScheduledStartStopEC2StopEC2LambdaB82BC7B0",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduledStartStopEC2EC2StopEventAllowEventRuleMyTestStackScheduledStartStopEC2StopEC2LambdaB536C9E720F2EEAD": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2StopEC2LambdaB82BC7B0",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2EC2StopEvent1CF6ECA1",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduledStartStopEC2StartEC2Lambda896098AC": {
      "DependsOn": [
        "ScheduledStartStopEC2StartEC2LambdaServiceRoleDefaultPolicyF0561FE0",
        "ScheduledStartStopEC2StartEC2LambdaServiceRole4CBFA4C9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "aecd7c4d20ea2e32da66de56148b8b6abec782319fdb72e50fd5d1314c6b0d95.zip",
        },
        "Environment": {
          "Variables": {
            "EC2_INSTANCE_ID": {
              "Ref": "SampleEC243038771",
            },
          },
        },
        "Handler": "start_ec2.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2StartEC2LambdaServiceRole4CBFA4C9",
            "Arn",
          ],
        },
        "Runtime": "python3.10",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduledStartStopEC2StartEC2LambdaServiceRole4CBFA4C9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledStartStopEC2StartEC2LambdaServiceRoleDefaultPolicyF0561FE0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ec2:Start*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledStartStopEC2StartEC2LambdaServiceRoleDefaultPolicyF0561FE0",
        "Roles": [
          {
            "Ref": "ScheduledStartStopEC2StartEC2LambdaServiceRole4CBFA4C9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduledStartStopEC2StopEC2LambdaB82BC7B0": {
      "DependsOn": [
        "ScheduledStartStopEC2StopEC2LambdaServiceRoleDefaultPolicy80FEB5B5",
        "ScheduledStartStopEC2StopEC2LambdaServiceRole2581D890",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "aecd7c4d20ea2e32da66de56148b8b6abec782319fdb72e50fd5d1314c6b0d95.zip",
        },
        "Environment": {
          "Variables": {
            "EC2_INSTANCE_ID": {
              "Ref": "SampleEC243038771",
            },
          },
        },
        "Handler": "stop_ec2.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "ScheduledStartStopEC2StopEC2LambdaServiceRole2581D890",
            "Arn",
          ],
        },
        "Runtime": "python3.10",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduledStartStopEC2StopEC2LambdaServiceRole2581D890": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledStartStopEC2StopEC2LambdaServiceRoleDefaultPolicy80FEB5B5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ec2:Stop*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledStartStopEC2StopEC2LambdaServiceRoleDefaultPolicy80FEB5B5",
        "Roles": [
          {
            "Ref": "ScheduledStartStopEC2StopEC2LambdaServiceRole2581D890",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
